pytorch_lightning/core/datamodule.py

```diff
@@ -86,6 +89,21 @@ def wrapped_fn(*args, **kwargs):
             if stage == "test" or stage is None:
                 obj._has_setup_test = True
 
+        # If calling teardown, we check the stage and assign stage-specific bool args
+        if fn.__name__ == "teardown":
+
+            # Get stage either by grabbing from args or checking kwargs.
+            # If not provided, set call status of 'fit' and 'test' to True.
+            # We do this so __attach_datamodule in trainer.py doesn't mistakenly call teardown('test') on trainer.test()
+            stage = args[1] if len(args) > 1 else kwargs.get("stage", None)
+
+            if stage == "fit" or stage is None:
+                obj._has_teardown_fit = True
+
+            if stage == "test" or stage is None:
+                obj._has_teardown_test = True
+
+
```

Comment:
I find this hook a bit wierd. fn should be called and then we set the variable to the model.

```
       outputs =  fn(*args, **kwargs)

        # If calling setup, we check the stage and assign stage-specific bool args
        if fn.__name__ == "setup":

            # Get stage either by grabbing from args or checking kwargs.
            # If not provided, set call status of 'fit' and 'test' to True.
            # We do this so __attach_datamodule in trainer.py doesn't mistakenly call setup('test') on trainer.test()
            stage = args[1] if len(args) > 1 else kwargs.get("stage", None)

            if stage == "fit" or stage is None:
                obj._has_setup_fit = True

            if stage == "test" or stage is None:
                obj._has_setup_test = True

        # If calling teardown, we check the stage and assign stage-specific bool args
        if fn.__name__ == "teardown":

            # Get stage either by grabbing from args or checking kwargs.
            # If not provided, set call status of 'fit' and 'test' to True.
            # We do this so __attach_datamodule in trainer.py doesn't mistakenly call teardown('test') on trainer.test()
            stage = args[1] if len(args) > 1 else kwargs.get("stage", None)

            if stage == "fit" or stage is None:
                obj._has_teardown_fit = True

            if stage == "test" or stage is None:
                obj._has_teardown_test = True


        if fn.__name__ == "prepare_data":
            obj._has_prepared_data = True

         return outputs

```

Reply:
@tchaton i agree, this current setup feels super hacky. i also don't like the duplicate code in trainer.py that's being added here either. the data module wrapper above seems like the source of this complication. this PR is actually broken since `teardown` isn't included in the wrapper alongside `setup`.